src/crimson/os/seastore/seastore_types.h

```diff
@@ -766,9 +766,25 @@ struct journal_seq_t {
     denc(v.offset, p);
     DENC_FINISH(p);
   }
+
+  // the comparisons ignore the segment-id
+  int cmp(const journal_seq_t& other) const {
+    if (segment_seq > other.segment_seq) {
+      return 1;
+    } else if (segment_seq < other.segment_seq) {
+      return -1;
+    }
+    int seg_off = offset.as_seg_paddr().get_segment_off();
+    int o_seg_off = other.offset.as_seg_paddr().get_segment_off();
+    return seg_off - o_seg_off;
+  }
+  bool operator==(const journal_seq_t& o) const { return cmp(o) == 0; }
+  bool operator!=(const journal_seq_t& o) const { return cmp(o) != 0; }
+  bool operator<(const journal_seq_t& o) const { return cmp(o) < 0; }
+  bool operator<=(const journal_seq_t& o) const { return cmp(o) <= 0; }
+  bool operator>(const journal_seq_t& o) const { return cmp(o) > 0; }
+  bool operator>=(const journal_seq_t& o) const { return cmp(o) >= 0; }
```

Comment:
Where are we doing this comparison?  I'm a bit leery of a tuple comparison that ignores one of the elements.  It may be correct in the case you are fixing, but I worry that future users will be surprised by this behavior.  I think other than this, we could:
- remove segment_id from journal_seq_t
- add a different journal_seq_t::cmp_seq_offset() static method with this behavior and update the caller
- change the comparison from (seq, id, offset) to (seq, offset, id) rather than to (seq, offset)

Reply:
> change the comparison from (seq, id, offset) to (seq, offset, id) rather than to (seq, offset)

See https://github.com/ceph/ceph/pull/46158/commits/4a7fe813b864d1a0122814a306f4781be531b7f4, I think this is more appropriate as most of the users of `journal_seq_t` needs the physical location from segment-id, such as journal-tail, scan_valid_records_cursor and write_result_t.

> Where are we doing this comparison?

As far as I know:
https://github.com/ceph/ceph/blob/75a9cbbc74ccab9a990a38f941c2f35c8570d76f/src/crimson/os/seastore/cache.cc#L1404-L1406
https://github.com/ceph/ceph/blob/75a9cbbc74ccab9a990a38f941c2f35c8570d76f/src/crimson/os/seastore/segment_cleaner.h#L1050
https://github.com/ceph/ceph/blob/75a9cbbc74ccab9a990a38f941c2f35c8570d76f/src/crimson/os/seastore/segment_cleaner.h#L1050
https://github.com/ceph/ceph/blob/75a9cbbc74ccab9a990a38f941c2f35c8570d76f/src/crimson/os/seastore/segment_cleaner.cc#L481-L514
https://github.com/ceph/ceph/blob/75a9cbbc74ccab9a990a38f941c2f35c8570d76f/src/crimson/os/seastore/seastore_types.cc#L648-L662